### DISCUSSION TITLE
Cecil will not work behind a proxy server?

### DISCUSSION THREAD

**Discussion by zemian** [Cecilapp/Cecil#1337] [Q&A] [answered]
Hi there, I am trying out Cecil, but the "serve" command hangs if I am behind a company proxy server. But it works fine if I am out of the private network. Is Cecil doing something special behind the scene that requires certain public hosts connection?

**Comment by ArnaudLigny** [comment]
Hello @zemian,

By default the _serve_ command run a local server available at `http://localhost:8000` ( `localhost` is a bind of `127.0.0.1`).

I suppose your company's proxy doesn't allows `localhost` alias and/or the `8000` port.

So, you can change the host and the port used by Cecil with the following options:
- host: `--host=X.X.X.X`
- port: `--port=XXXX`

e.g.: `cecil serve --host=192.168.1.15 --port=8080`.

Docs: https://cecil.app/documentation/commands/#serve.

I hope I've helped you ! 😃

**Comment by zemian** [comment]
Hey @ArnaudLigny thanks for the reply! No, my company proxy does not block `localhost`. I can run other applications on `localhost:8000` just fine. Example, `php -S localhost:8000` works without any issue. Only when running `cecil serve` command it hangs.

Running with `-v` flag I see it stuck here...

```bash
$ php cecil.phar serve -v
Building website...
Path: /myproject
Config: /myproject/config.yml
Cache: /myproject/.cache
1. Loading content
2. Creating pages
3. Converting pages
```

**Reply by ArnaudLigny** [reply]
Hello, If an error is not correctly catched you can get a more verbose log with the following options : `-vv` or `-vvv` (debug mode).

**Comment by zemian** [comment]
I think I figured it out why it hangs. I read the Cecil docs that it will parse images and create few responsive versions of it. The sample `content/index.md` has this line:

```markdown
![Octocat](https://github.githubassets.com/images/icons/emoji/octocat.png)
```

Removing this line works! So I am guessing `https://github.githubassets.com` is blocked by my company proxy. I hope this helps some future users!

**Reply by ArnaudLigny** [reply]
Cecil do not create responsives images by default (option must be enabled in configuration, see https://cecil.app/documentation/configuration/#assets) but it downloads remote images and put them in cache.

Your case (behind a proxy) is interesting and I think I should code a better error handling with remotes images.

**Comment by zemian** [comment]
Hi @ArnaudLigny , actually this brings me to next question. The docs https://cecil.app/documentation/configuration#assets says images responsive creation is turn off by default, and yet it's still generating it. In fact I explicitly set it in `config.yml` file to off, and I still see it gets generated. Is this a bug, or is there another way to turn it off?

```yaml
images:
  enabled: false
  optimize:
    enabled: false
  responsive:
    enabled: false
```

**Reply by ArnaudLigny** [reply]
I confirms Cecil don't create responsive images by default. So it's strange ^^

Could you share a public repository of your project?

**Reply by zemian** [reply]
Hey @ArnaudLigny I don't have a repo. I simply followed your docs on getting started guide. Here are the steps:

```
# Download cecil.phar
$ php cecil.phar new:site cecil-test-site
Creating a new website...
Done!
$ php cecil.phar build -vvv cecil-test-site
...
find cecil-test-site/_site

cecil-test-site/_site
cecil-test-site/_site/index.html
cecil-test-site/_site/favicon.c1af8a129a0cde81f9b94ffbf452e10b.png
cecil-test-site/_site/404.html
cecil-test-site/_site/atom.xml
cecil-test-site/_site/sitemap.xml
cecil-test-site/_site/_debug_twig_profile.html
cecil-test-site/_site/robots.txt
cecil-test-site/_site/another-page
cecil-test-site/_site/another-page/index.html
cecil-test-site/_site/assets
cecil-test-site/_site/assets/github.githubassets.com
cecil-test-site/_site/assets/github.githubassets.com/images
cecil-test-site/_site/assets/github.githubassets.com/images/icons
cecil-test-site/_site/assets/github.githubassets.com/images/icons/emoji
cecil-test-site/_site/assets/github.githubassets.com/images/icons/emoji/octocat.1b27964ad692434f271a1dcc82277351.png
cecil-test-site/_site/assets/thumbnails
cecil-test-site/_site/assets/thumbnails/57
cecil-test-site/_site/assets/thumbnails/57/favicon.c1af8a129a0cde81f9b94ffbf452e10b.png
cecil-test-site/_site/assets/thumbnails/32
cecil-test-site/_site/assets/thumbnails/32/favicon.c1af8a129a0cde81f9b94ffbf452e10b.png
cecil-test-site/_site/assets/thumbnails/192
cecil-test-site/_site/assets/thumbnails/192/favicon.c1af8a129a0cde81f9b94ffbf452e10b.png
cecil-test-site/_site/assets/thumbnails/1200
cecil-test-site/_site/assets/thumbnails/1200/cecil-card.c0bc77e003455fafeb15f3072e857d85.png
cecil-test-site/_site/assets/thumbnails/228
cecil-test-site/_site/assets/thumbnails/228/favicon.c1af8a129a0cde81f9b94ffbf452e10b.png
cecil-test-site/_site/assets/thumbnails/180
cecil-test-site/_site/assets/thumbnails/180/favicon.c1af8a129a0cde81f9b94ffbf452e10b.png
cecil-test-site/_site/assets/thumbnails/120
cecil-test-site/_site/assets/thumbnails/120/favicon.c1af8a129a0cde81f9b94ffbf452e10b.png
cecil-test-site/_site/assets/thumbnails/128
cecil-test-site/_site/assets/thumbnails/128/favicon.c1af8a129a0cde81f9b94ffbf452e10b.png
cecil-test-site/_site/assets/thumbnails/196
cecil-test-site/_site/assets/thumbnails/196/favicon.c1af8a129a0cde81f9b94ffbf452e10b.png
cecil-test-site/_site/assets/thumbnails/96
cecil-test-site/_site/assets/thumbnails/96/favicon.c1af8a129a0cde81f9b94ffbf452e10b.png
cecil-test-site/_site/assets/thumbnails/152
cecil-test-site/_site/assets/thumbnails/152/favicon.c1af8a129a0cde81f9b94ffbf452e10b.png
cecil-test-site/_site/assets/thumbnails/76
cecil-test-site/_site/assets/thumbnails/76/favicon.c1af8a129a0cde81f9b94ffbf452e10b.png
cecil-test-site/_site/cecil-card.png
```

I see the "octocat" image there. It's not creating various responsive copies, but it still downloaded the image. That's the issue that I am facing. How can I turn this off and not download it?

I see "thumbnails" been generated for various responsive image size format. Hence I thought it's doing all the image, but now I look closer, it's just the favicon image. But even for that, how can I turn it off?

Anyway, if you say Cecil is downloading the image for cache only, then it is still a problem right? The problem also exists when I run "cecil build", in this case does it still generate cache? Why build requires cache? 

My point is, this hanging issue is not a pleasant one for a first time user. I just want to generate a site with local files. I don't expect the static generator tool to go out to internet and download stuff while I am building a site. I hope this feedback helps. :)

**Reply by ArnaudLigny** [reply]
> I see "thumbnails" been generated for various responsive image size format. Hence I thought it's doing all the image, but now I look closer, it's just the favicon image. But even for that, how can I turn it off?

Haaaa, I see why: this is because the default site (skeleton) includes the **metatags** template. This partial template is an SEO helper that is able to (among others) generate deferent sizes of the same `favicon.png` image.
You can disable it, see the documentation: https://cecil.app/documentation/configuration#metatags-configuration.

Cecil don't require internet access, and fetch remote file(s) only if you use one. This is the case `octocat.png` . So you can remove it from from `content/index.md` (https://github.com/Cecilapp/Cecil/blob/master/resources/skeleton/content/index.md).

**Reply by zemian** [reply]
I see. Thanks for the `metatags` config docs link!

So in the case of `octocat.png`, if I want to keep it in the `content/index.md`, how do I make Cecil not to download the image into `assets` folder?

**Reply by ArnaudLigny** [reply]
> how do I make Cecil not to download the image into assets folder?

A remote image is not stored in the `assets` directory of your project, but in the temporary directory `.cache/assets`.

The cache is preserved between each build except if your run the build command in debug mode (with `-vvv` the cache is cleared) or if you run the commands `cecil clear`, `cecil cache:clear`.

And sorry: there is no option to disable the fetch process of remote image.

**Reply by zemian** [reply]
Based on the output I capture above, the image is in the generated folder `cecil-test-site/_site/assets/github.githubassets.com/images/icons/emoji/octocat.1b27964ad692434f271a1dcc82277351.png`, not just in cache folder! And I don't need this.

Hm.. and there is no option to turn this off! Okay. It might be just me, but I find this behavior odd.

But hey thanks for your patient and clarifying this to me! Have a good day!

**Reply by ArnaudLigny** [reply]
Indeed, I was not very clear, sorry 😅

You are right: currently a remote asset (image or other file type) is systematically downloaded and stored in the output (`_site`) in order to facilitate any handling (resize, compression, etc.) and to be sure the built website can be browse offline.

Maybe I should add an option to disable that.

**Reply by zemian** [reply]
Yes, that would be great to have as option. I have help created a feature request here to track it: https://github.com/Cecilapp/Cecil/issues/1338

**Comment by ArnaudLigny** [comment, marked as answer]
It should be fixed with the feature mentioned in PR https://github.com/Cecilapp/Cecil/pull/1342.
